hpx/util/detail/basic_function.hpp

```diff
@@ -119,24 +126,29 @@ namespace hpx { namespace util { namespace detail
 
         void reset() noexcept
         {
-            if (!empty())
+            if (object != nullptr)
             {
-                vptr->delete_(object);
+                vptr->delete_(object, function_storage_size);
 
                 vptr = detail::get_empty_function_vtable<VTable>();
-                vtable::default_construct<empty_function>(object);
+                object = nullptr;
             }
         }
 
         void swap(function_base& f) noexcept
         {
             std::swap(vptr, f.vptr);
-            std::swap(object, f.object); // swap
+            std::swap(object, f.object);
+            std::swap(storage, f.storage);
+            if (object == &f.storage)
+                object = &storage;
+            if (f.object == &storage)
+                f.object = &f.storage;
```

Comment:
What's the purpose of these assignments?

Reply:
The object pointer may point to embedded storage, in which case it would be pointing to the wrong embedded storage after the swap. If the object pointer was pointing to the other's embedded storage before, it has to point to its own embedded storage now.